packages/real-time-client/src/client.ts

```diff
@@ -0,0 +1,270 @@
+    if (!this.socket) return undefined;
+    return {
+      [WebSocket.CONNECTING]: 'connecting' as const,
+      [WebSocket.OPEN]: 'open' as const,
+      [WebSocket.CLOSING]: 'closing' as const,
+      [WebSocket.CLOSED]: 'closed' as const,
+    }[this.socket.readyState];
+  }
+
+  // Track the last AudioAdded sequence number, used when stopping transcription to avoid missing audio
+  // https://docs.speechmatics.com/rt-api-ref#audioadded
+  private lastAudioAddedSeqNo = 0;
+
+  private async connect(jwt: string) {
+    return new Promise<void>((resolve, reject) => {
+      const url = new URL(this.url);
+      url.searchParams.append('jwt', jwt);
+      if (this.appId) {
+        url.searchParams.append('sm-app', this.appId);
+      }
+
+      this.socket = new WebSocket(url.toString());
+      this.dispatchTypedEvent(
+        'socketStateChange',
+        new Event('socketStateChange'),
+      );
+
+      this.socket.addEventListener(
+        'open',
+        () => {
+          resolve();
+        },
+        { once: true },
+      );
+
+      this.socket.addEventListener('error', (error) => {
+        console.error(error);
+        this.dispatchTypedEvent(
+          'socketStateChange',
+          new Event('socketStateChange'),
+        );
+        // In case the above hasn't resolved, we can reject here rather than waiting
+        // If the above has resolved, this will be ignored
+        reject(error);
+      });
+
+      this.socket.addEventListener('close', () => {
+        this.dispatchTypedEvent(
+          'socketStateChange',
+          new Event('socketStateChange'),
+        );
+      });
+
+      this.socket.addEventListener('message', (socketMessage) => {
+        // Turn message into one of the union type of events
+        // Pass it to this event target
+        const data = JSON.parse(socketMessage.data) as unknown;
+        if (!dataIsRealtimeTranscriptionMessage(data)) {
+          console.warn(
+            'message does not look like a valid message: ',
+            JSON.stringify(data),
+          );
+          return;
+        }
+
+        if (data.message === 'Error') {
+          console.error(data);
+        }
+        if (data.message === 'AudioAdded') {
+          this.lastAudioAddedSeqNo = data.seq_no;
+        }
+
+        this.dispatchTypedEvent(
+          'receiveMessage',
+          new MessageEvent('receiveMessage', { data }),
+        );
+      });
+    });
+  }
+
+  sendMessage(message: RealtimeClientMessage) {
+    if (!this.socket) {
+      console.warn('Client socket not initialized');
+      return;
+    }
+    this.socket.send(JSON.stringify(message));
+    this.dispatchTypedEvent(
+      'sendMessage',
+      new MessageEvent('sendMessage', { data: message }),
+    );
+  }
+
+  sendAudio(data: Blob | ArrayBufferLike | string) {
+    if (!this.socket || this.socket.readyState !== this.socket.OPEN) {
+      console.warn('Socket not ready to receive audio, will not pass data');
```

Comment:
There's a lot of console warns in this package. Maybe we could cut back on them, or think about whether they should count as warns or throw errors? e.g. calling send audio or send message on a client should perhaps throw an error. Not sure what your thoughts are on this - maybe also not something to discuss in this major structural refactor!